testonly/hammer/hammer.go

```diff
@@ -284,6 +285,75 @@ func newWorker(cfg *MapConfig, prng *rand.Rand) *mapWorker {
 	}
 }
 
+// readWorker performs read-only operations on a fixed map.
+type readWorker struct {
+	*mapWorker
+
+	validReadOps   *validReadOps
+	invalidReadOps *invalidReadOps
+}
+
+func newReadWorker(s *hammerState, idx int) *readWorker {
+	return &readWorker{
+		mapWorker: newWorker(s.cfg, rand.New(rand.NewSource(int64(idx)))),
+
+		validReadOps:   s.validReadOps,
+		invalidReadOps: s.invalidReadOps,
+	}
+}
+
+func (w *readWorker) loopUntilDone(ctx context.Context, done <-chan struct{}) error {
+	for {
+		select {
+		case <-done:
+			return nil
+		default:
+		}
+		if err := w.validReadOps.getLeavesRev(ctx, w.prng); err != nil {
+			if _, ok := err.(errSkip); ok {
+				continue
+			}
+			return err
+		}
+	}
+}
+
+// writeWorker performs mutation operations on a fixed map.
+type writeWorker struct {
+	*mapWorker
+
+	operations uint64
+
+	// TODO(mhutchinson): Remove hammerState from here - it allows access to global info
+	// which makes reasoning about the behaviour difficult.
+	s *hammerState
+}
+
+func newWriteWorker(s *hammerState) *writeWorker {
+	return &writeWorker{
+		mapWorker: newWorker(s.cfg, rand.New(s.cfg.RandSource)),
+
+		operations: s.cfg.Operations,
+		s:          s,
+	}
+}
+
+func (w *writeWorker) loop(ctx context.Context, done <-chan struct{}) (uint64, error) {
+	count := uint64(0)
+
+	for ; count < w.operations; count++ {
+		select {
+		case <-done:
+			return count, nil
+		default:
+		}
+		if err := w.retryOneOp(ctx, w.s); err != nil {
+			return count, err
+		}
+	}
+	return count, nil
```

Comment:
It seems that we get to this line only when `count == w.operations`. How about `return w.operations, nil` and inlining `count := uint64(0)` to the `for` statement?

Reply:
I think I prefer it the way it is. This way is clear - I like the consistency that every return is `return count, ...` - and I'm not too concerned with holding onto a single uint64 for an extra beat.